src/compiler.rkt

```diff
@@ -2,31 +2,98 @@
 
 (require math/bigfloat rival)
 (require "syntax/syntax.rkt" "syntax/types.rkt"
-         "common.rkt" "timeline.rkt" "float.rkt")
+         "common.rkt" "timeline.rkt" "float.rkt" "config.rkt")
 
 (provide compile-specs compile-spec compile-progs compile-prog)
 
+;; Function calculates a precision for the operation
+;;    with respect to the given extra-precision (exponents from a previous run)
+(define (operator-precision extra-precision)
+  (min (*max-mpfr-prec*)
+       (+ extra-precision
+          (*ground-truth-extra-bits*)
+          (bf-precision))))
+
+(define (true-exponent x)
+  (+ (bigfloat-exponent x) (bigfloat-precision x)))
+
+(define (unbox-prec x)
+    (if (box? x)
+        (unbox x)
+        0))
+
 ;; Interpreter taking a narrow IR
 ;; ```
 ;; <prog> ::= #(<instr> ..+)
 ;; <instr> ::= '(<op-procedure> <index> ...)
+;; <op-procedure> ::= | #(<operation> <extra-precision> <exponents-checkpoint>) - tuned op
+;;                    | #(<operation> <extra-precision>) - regular op
+;;                    | <operation> - const
 ;; ```
 ;; where <index> refers to a previous virtual register.
 ;; Must also provide the input variables for the program(s)
 ;; as well as the indices of the roots to extract.
-(define (make-progs-interpreter vars ivec roots)
+;; name ::= 'fl or 'ival
+(define (make-progs-interpreter name vars ivec roots)
   (define vreg-count (+ (length vars) (vector-length ivec)))
   (define vregs (make-vector vreg-count))
-  (λ args
-    (for ([arg (in-list args)] [n (in-naturals)])
-      (vector-set! vregs n arg))
-    (for ([instr (in-vector ivec)] [n (in-naturals (length vars))])
-      (define srcs
-        (for/list ([idx (in-list (cdr instr))])
-          (vector-ref vregs idx)))
-      (vector-set! vregs n (apply (car instr) srcs)))
-    (for/list ([root (in-list roots)])
-      (vector-ref vregs root))))
+
+  
+  (define (tuning-filter instr)
+    (if (vector? (car instr))
+        (if (member
+             (vector-ref (car instr) 0)
+             (list ival-sin ival-cos ival-tan))
+            #t
+            #f)
+        #f))
+  
+  (define tuning-ivec
+    (if (equal? name 'ival)
+        (vector-filter tuning-filter ivec)
+        '()))
+
+  (if (equal? name 'ival)
+    (λ args
+      ;; remove all the exponent values we assigned previously when a new point comes
+      (when (equal? (bf-precision) (*starting-prec*))
+        (for ([instr (in-vector tuning-ivec)])
+          (set-box! (vector-ref (car instr) 2) 0)))
+    
+      (for ([arg (in-list args)] [n (in-naturals)])
+        (vector-set! vregs n arg))
+      (for ([instr (in-vector ivec)] [n (in-naturals (length vars))])
+        (define srcs
+          (for/list ([idx (in-list (cdr instr))])
+            (vector-ref vregs idx)))
+        
+        (match (car instr)
+          [(vector op extra-precision exponents-checkpoint)  ; current op is sin/cos/tan
```

Comment:
now that there's only one branch this can be a `match-define`, and the `let` can be a `define`.